### ISSUE TITLE
Question about missingness

### ISSUE THREAD

**Issue by psinger** [shap/shap#175]
Hope this is the right place to ask a question like this, but I am trying to get my head around the missing property. In one of the papers it is written that

> If the simplified inputs represent feature presence, then missingness requires features missing in the original input to have no impact.

What exactly does missing mean in this context? If a feature is zero, it of course also can have an impact on the model, and actually shap values also are non-zero for zero-input features.

Thanks in advance for an explanation!

**Comment by slundberg** [comment]
Yes, this is the right place. The missingness property is really just a minor book-keeping property to close a loop hole. It is required since local accuracy is specified as a linear model and `x'` could in theory have some zero entries (meaning the input is already missing as defined by `h_x`). These zero entries mean that local accuracy would still hold no matter what phi values correspond to those entries, so to have a unique solution we need to constrain them to be 0 (which is what we want since they are missing already and so have no impact). In practice for SHAP we will never consider a feature to already be perfectly missing unless that feature's value is constant over the whole background dataset.

**Comment by psinger** [comment]
Thanks. How is `h_x` actually defined and what role does it play in calculating the shap values?

**Comment by slundberg** [comment]
`h_x` is used to connect SHAP with LIME, for SHAP it essentially replaces each missing input with a missing at random value indicator (meaning the variable should be integrated out). But this is just a long way of saying that f(h_x(z')) = E[f(x) | x_S]` where `S` is the set of non-zero entries in the binary vector `z'`.

**Comment by liuyanguu** [comment]
I happen to have the same question and I don't really know how is the SHAP value produced for missing (`NA`) feature values... I compared the SHAP values for the NA feature values and non-NA feature values, the SHAP values for NA feature values are roughly 10-times smaller on magnitude. I guess the reason traces back to `xgboost` being a sparsity aware algorithm?

**Comment by slundberg** [comment]
@liuyanguu NA in XGBoost is just routed down one of the two branches whenever you are splitting on that feature. So as far as XGBoost is concerned, NA doesn't mean missing, it is just a special value to split on (which you might use to represent missing). Since NA is really just a special indicator, it does not always mean missing at random (which is what SHAP means by missing). When Tree SHAP means missing, it means integrating the variable out by following both branches of the tree splits weighted by the number of training samples that went each way.

If NA does mark samples that are nearly missing at random, then their SHAP values will be very small, since the result of following the NA path will be similar to the expected output if you integrated over all paths accessible by toggling that feature.

**Comment by IvanUkhov** [comment]
@slundberg, as you mentioned, XGBoost has no problem with missing values, which, in my case, are encoded as NaNs. I wonder if Shap can also gracefully handle examples containing such values.

I see some NaNs in the output of `shap_values`. Is it expected? If not, what would you suggest doing?

**Comment by slundberg** [comment]
@IvanUkhov that is not expected. I need to debug that. I think it is happening because of numerical precision issues on deep trees. See #152

**Comment by xervanyo** [comment]
Hello,
I am currently trying to simulate the Tree SHAP missingness of a variable, as @slundberg mentionned "integrating the variable out by following both branches of the tree splits weighted by the number of training samples that went each way." 

Do you have an easy way to program that, because I am really struggling. 

Thank you

**Comment by slundberg** [comment]
@xervanyo Here is an example: https://github.com/slundberg/shap/blob/04e7e35bde07484243a783fee5f0c32d4d41571e/tests/explainers/test_tree.py#L72-L96

**Comment by zyxue** [comment]
A follow-up question, please. 

Is missingness the same thing as null effects?

**Comment by slundberg** [comment]
Plugging null into the model is not the same as how SHAP computes the effect of a missing feature (until you provide a background dataset of all null values).

**Comment by ManuelZ** [comment]
@slundberg, In the "Explainable AI for Trees" paper, "missingness" sounds exactly like the "dummy players" axiom of the Shapley values. But in the "A Unified Approach to Interpreting Model Predictions" paper, "missingness" sounds different, it sounds like this: if in the simplified input there is a missing feature (with respect to the original input), then the Shapley value of such missing feature should be zero (which sounds straightforward: if it's not present in the coalition, then it contributes nothing).

Is the way I interpret it in the original paper correct?

**Comment by AneeshBose** [comment]
Hi @slundberg, I was using the SHAP library to explain examples trained on xgboost and got non-zero shap values for features with data as 0, which, as per my understanding, is the feature's contribution to f(x) based on its interaction with other features. However, to distinguish missing values that are imputed with actual zero values, I retrained the model by adding -1 to 0 values that can never actually be 0 in reality. However, the shap value is changing quite a bit. Ideally, I would want these shap values to be as close to 0 as they should not have much influence on the local explanation. Is there any best practice to be followed here?

**Comment by GolnooshBabaei** [comment]
@slundberg I hope my question is right. I am thinking about changing the part of the kernelshap in which a value (I guess by default mean) is calculated based on the given background dataset for the missing feature in the coallition. Unfortunately, I couldn't find how it is computed through the code. Could you please tell me in which function it is computed?

**Comment by cmougan** [comment]
I believed that property is also relevant if the feature has no relation to the target. E.g. a random feature.
Then the value will always be 0. No matter the coalition. 

Perhaps I am mixing the Dummy properties of the  Shapley values

**Comment by eduardohorta** [comment]
> `h_x` is used to connect SHAP with LIME, for SHAP it essentially replaces each missing input with a missing at random value indicator (meaning the variable should be integrated out). But this is just a long way of saying that f(h_x(z')) = E[f(x) | x_S]`where`S`is the set of non-zero entries in the binary vector`z'`.

Typically we have $f(x) = f(x,{\rm T}),$ where ${\rm T}$ is the training data. From the theoretical viewpoint, should I interpret $E[f(X,{\rm T})\ |\  X_S=x_s]$ as $\int f(u,T)\ P_{X|X_S=x_S}({\rm d}u)$ or as $\int f(u,t)\ P_{X,{\rm T}|X_S=x_S}({\rm d}u,{\rm d}t)$?
